### PR TITLE
[Pal/Linux-SGX] Add MAA to RA-TLS and SecretProv libs

### PR DESCRIPTION
**Pull request by dimakuv**
## Description of the changes 

Microsoft Azure Attestation (MAA) is a new attestation scheme available in the MS Azure public cloud. It uses the classic DCAP SGX quotes but replaces the DCAP/EPID quote-verification software infrastructure with the new MAA attestation provider infrastructure. In particular, the remote user is supposed to send the SGX quote in the HTTPS REST attestation request to the MAA attestation provider and receive back the JSON Web Token (JWT) that contains claims about the attesting SGX enclave.

This commit introduces new `_maa` versions of the RA-TLS and Secret Provisioning verification libraries. The documentation and the `ra-tls-mbedtls` and `ra-tls-secret-prov` examples are updated.

For design, see https://github.com/gramineproject/gramine/issues/626.

This PR does *not* depend on any other PRs, but as part of this work I detected and fixed some attestation-related issues:
- https://github.com/gramineproject/gramine/pull/638
- https://github.com/gramineproject/gramine/pull/643
- https://github.com/gramineproject/gramine/pull/645
- https://github.com/gramineproject/gramine/pull/647

## How to test this PR? 

Manually verified on a MS Azure CC VM, with the `https://sharedcus.cus.attest.azure.net` default attestation provider. See the READMEs of `ra-tls-mbedtls` and `ra-tls-secret-prov` examples.

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/652)

### CODE REVIEW

File: CI-Examples/ra-tls-mbedtls/src/client.c [hunk 1]
```diff
@@ -200,6 +200,21 @@ int main(int argc, char** argv) {
...
+            mbedtls_printf("%s\n", dlerror());
+            mbedtls_printf("User requested RA-TLS verification with MAA but cannot find helper"
+                           " libsgx_urts.so lib\n");
+            return 1;
+        }
+
+        ra_tls_verify_lib = dlopen("libra_tls_verify_maa.so", RTLD_LAZY);
+        if (!ra_tls_verify_lib) {
+            mbedtls_printf("%s\n", dlerror());
+            mbedtls_printf("User requested RA-TLS verification with MAA but cannot find lib\n");
```
Comment (bigdata-memory) [inline CI-Examples/ra-tls-mbedtls/src/client.c:215]: Could be more specific in this context.
For example. `"User requested RA-TLS verification with MAA but cannot find helper libra_tls_verify_maa.so lib.\n"`

File: CI-Examples/ra-tls-mbedtls/src/client.c [hunk 2]
```diff
@@ -200,6 +200,21 @@ int main(int argc, char** argv) {
...
+    } else if (!strcmp(argv[1], "maa")) {
+        void* helper_sgx_urts_lib = dlopen("libsgx_urts.so", RTLD_NOW | RTLD_GLOBAL);
+        if (!helper_sgx_urts_lib) {
+            mbedtls_printf("%s\n", dlerror());
+            mbedtls_printf("User requested RA-TLS verification with MAA but cannot find helper"
+                           " libsgx_urts.so lib\n");
+            return 1;
+        }
+
+        ra_tls_verify_lib = dlopen("libra_tls_verify_maa.so", RTLD_LAZY);
```
Comment (bigdata-memory) [inline CI-Examples/ra-tls-mbedtls/src/client.c:212]: Suggest using a `const char*` to hold the name of this external library for reuse in message if possible.

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/652)* status: 0 of 22 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @bigdata-memory)

___
*[`CI-Examples/ra-tls-mbedtls/src/client.c` line 212 at r1](https://reviewable.io/reviews/gramineproject/gramine/652#-N4XOP32JTuAZo5JPcV2-r1-212:-NELTAxz8zmqDLLjERVy:b8p55w7) ([raw file](https://github.com/gramineproject/gramine/blob/3abc5f03e8b048cf5f45c9b8b5ec5d65965c23f2/CI-Examples/ra-tls-mbedtls/src/client.c#L212)):*
<details><summary><i>Previously, bigdata-memory (Gordon King) wrote…</i></summary><blockquote>

Suggest using a `const char*` to hold the name of this external library for reuse in message if possible.
</blockquote></details>

Don't see a need.
___
*[`CI-Examples/ra-tls-mbedtls/src/client.c` line 215 at r1](https://reviewable.io/reviews/gramineproject/gramine/652#-N4XOP32JTuAZo5JPcV2-r1-215:-NELTCRG6eG3OYVcMp5B:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/3abc5f03e8b048cf5f45c9b8b5ec5d65965c23f2/CI-Examples/ra-tls-mbedtls/src/client.c#L215)):*
<details><summary><i>Previously, bigdata-memory (Gordon King) wrote…</i></summary><blockquote>

Could be more specific in this context.
For example. `"User requested RA-TLS verification with MAA but cannot find helper libra_tls_verify_maa.so lib.\n"`
</blockquote></details>

Done.

**Comment by xiangyuT** [comment]
Hi @dimakuv, I want to know when will this PR be merged and before it's merged, if we can and how we can do MAA in Gramine?

**Comment by dimakuv** [comment]
@xiangyuT 

> I want to know when will this PR be merged...

Hard to say, but in upcoming 1-2 months I think. It depends on the review process. Which is a good opportunity to ping @mkow :)

> ...before it's merged, if we can and how we can do MAA in Gramine?

You can use this PR (this branch) to test MAA with Gramine. There is documentation included in this PR. Or maybe you meant something else?

**Comment by xiangyuT** [comment]
@dimakuv
Got it. I will try this branch to test MAA. Thanks!

**Review by mkow — COMMENTED** [review]
> Which is a good opportunity to ping @mkow :)

ITL is still waiting to get an Azure instance from either Intel or MS, otherwise we can't even test this PR. Also, what about CI? ;)

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/652#-:-NGXmkfH3G8KAWvWa_zM:b61tqnb)* status: 0 of 22 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @bigdata-memory)

**Review by dimakuv — COMMENTED** [review]
@mkow I can give you access to my Azure VM with Ubuntu 20.04 for testing. Do you want that?

Regarding CI: I don't have a good idea. I also don't have a good idea how MAA can be split as some kind of "plugin" from the rest of RA-TLS code...

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/652#-:-NG_ib2j7JATaAJKZeri:byvjuwj)* status: 0 of 22 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @bigdata-memory)

**Comment by RodgerZhu** [comment]
Hi @dimakuv, any latest updates about this PR?
We are now working on Azure SGX instance with Gramine and this MAA requirement is blocking us.

**Comment by dimakuv** [comment]
@RodgerZhu Short answer is that this PR is pending on https://github.com/gramineproject/gramine/pull/1114.

I.e., Gramine first needs to have a proper plugin architecture for external attestation schemes (like MAA), and then I'll refactor this PR to become a plugin.

The timeline is uncertain for this feature...

**Pull request closed without merging**